### PR TITLE
Ignore errors parsing slicec JSON output

### LINKED ISSUE #4043: Slice.Tools JSON error
I am getting this error whenever I get a redefinition error in a Slice file:

```
    /Users/bernard/builds/icerpc-csharp/tools/IceRpc.Slice.Tools/IceRpc.Slice.Tools.targets(60,5): error : Error parsing JSON: 'F' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.. JSON content: Failed: Compilation failed with 3 error(s)
```

### PR DESCRIPTION
**Pull request by pepone**
Fix #4043

### CODE REVIEW

File: tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs [hunk 1]
```diff
@@ -109,27 +109,23 @@ protected override string GenerateFullPathToTool()
...
     {
+        Console.WriteLine(singleLine);
         if (messageImportance == MessageImportance.Low)
         {
-            try
+            // Messages from stdout
+            if (JSONParser.Parse<BuildTelemetry>(singleLine) is BuildTelemetry buildTelemetry)
             {
-                // Messages from stdout
                 var jsonDoc = System.Text.Json.JsonDocument.Parse(singleLine);
```
Comment (externl) [inline tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs:118]: Is this variable still needed?

**Review by externl — APPROVED** [review]

**Review by externl — COMMENTED** [review]
Should we first check that the line (whitespace stripped) begins and ends with `{` and `}` instead of just ignoring all Json exceptions. Otherwise it might not be obvious if we break something?

File: tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs [hunk 2]
```diff
@@ -109,27 +109,23 @@ protected override string GenerateFullPathToTool()
     /// </summary>
     protected override void LogEventsFromTextOutput(string singleLine, MessageImportance messageImportance)
     {
+        Console.WriteLine(singleLine);
```
[the remark this answers is not in the document]
Comment (pepone) [reply tools/IceRpc.Slice.Tools/SliceCCSharpTask.cs:112]: Good catch, that is a leftover!

**Comment by pepone** [comment]
> Should we first check that the line (whitespace stripped) begins and ends with { and } instead of just ignoring all Json exceptions. Otherwise it might not be obvious if we break something?

The plug-in should not code defensively against possible bugs in the compiler.

I think a better compromise here would be to emit all messages as JSON when `--diagnostic JSON` is set, then log any exceptions as fatal errors instead of eating them.

The compiler can emit `{summary: "Compilation failed with 1 error(s)" status: 1}` instead of `Failed: Compilation failed with 1 error(s)`

**Comment by InsertCreativityHere** [comment]
> The plug-in should not code defensively against possible bugs in the compiler.

Agreed. but no matter what we do, the compiler will sometimes emit non-json stuff, and I don't think we should eat and ignore them. For example, if you pass a bad CLI option, the compiler exits before it even has a chance to properly set json-diagnostics.

> I think a better compromise here would be to emit all messages as JSON

Actually, could we go even simpler, and just not emit this summary message at all?
JSON mode is basically only meant for tools/plug-ins. And if they're just going to ignore the message, why emit it?
All the information in the message is recoverable anyways. You already know how many errors there were, and if that number is greater than 0, then the status code is going to be 1.

**Comment by pepone** [comment]
> Actually, could we go even simpler, and just not emit this summary message at all?

Yes tools are just going to ignore it.

> Agreed. but no matter what we do, the compiler will sometimes emit non-json stuff,

Right, the ignoring of JSON was just to work around the summary message not being JSON formatted.

**Comment by InsertCreativityHere** [comment]
Cool, I'll open a PR for `slicec` to suppress the summary message when using JSON mode!
Just give me a couple minutes.

**Comment by InsertCreativityHere** [comment]
See: https://github.com/icerpc/slicec/pull/702

**Comment by InsertCreativityHere** [comment]
With the changes made to the Slice compiler, the original issue should be fixed now.

But I still think that this centralized parser is nicer, and worth doing. The only change I would recommend is emitting an error message when we hit an exception in the JSON parsing, instead of ignoring them.

**Pull request closed without merging**